### ISSUE TITLE
Easier weight sharing

### ISSUE THREAD

**Issue by rnett** [deeplearning4j/deeplearning4j#7408]
For implementing more complex multi-output layers, such as GANs or adversarial autoencoders (an autoencoder with a discriminator on the reconstruction), it would be nice to have an easier way to share weights between layers in different models.

As per #791, the current way to implement GANs is to train the discriminator, then copy the weights into the GAN, then train the GAN to fool the discriminator and copy the weights into the generator.  This is very clunky.

Contrast this with Keras's implantation of shared networks, where you simply create 3 models from the same layers, and can train each independently.

The main thing that makes this easier is that in Keras, if you use the same layer object in two different models, the weights are shared.

From what I've looked at, a possible implementation is to give every layer and vertex a "weight pool" object, that `paramTable` and `params` interact with, that is basically a wrapper for any parameters.  This can be created and shared by the layer configs, and passed down to the implementations.

Then you could have a method like `(config)Layer.copyAndShareWeights()` which returns a new config with the same weight pool object.

Also, you could have things like getting a subset of a network with shared weights.

It would most likely have to interact with memory workspaces, distributed training, and serialization (probably just index pools).

**Comment by rnett** [comment]
Another possible route is to cache the implementation layer created by `instantiate`.

**Comment by AlexDBlack** [comment]
See also https://github.com/deeplearning4j/deeplearning4j/issues/1785

> Another possible route is to cache the implementation layer created by instantiate.

That won't work, due to how SGD, updaters, model persistence etc are implemented internally.

> From what I've looked at, a possible implementation is to give every layer and vertex a "weight pool" object, that paramTable and params interact with, that is basically a wrapper for any parameters. This can be created and shared by the layer configs, and passed down to the implementations.

That's one possible design, not sure it would result in a clean API.

One API design we've had in mind for this in DL4J is something like ```.shareParameters("x", "y")``` to share the parameters for layers X and Y (which must be the same layer type, obviously). Such a method could be called as many times as necessary to create additional parameter sharing between layers. I like that from an API perspective because it's an unambiguous declaration of what the user intends to do.

(this is trivial in SameDiff fwiw, at last if you're making your layers directly - just use the weights/biases in multiple places and it'll work)

**Comment by AlexDBlack** [comment]
OK, so there's basically 2 issues here:
(1) User-facing API for network construction. That's the easy part.
(2) Back-end implementation - there's some challenges here, and it's why we don't have it already.

Those issues are all solveable, it's been an issue of finding the time to do it.

OK, some issues we'll run into here:

**Issue 1: Parameters and gradients in DL4J use a row vector, which is split up (views) for each layer.**
Number of parameters for the net is determined as the sum of the number of parameters for each layer.
As soon as you add parameter sharing: that assumption breaks. Same thing for updater state, and gradient view array.
See init() and initGradientsView() in MultiLayerNetwork/ComputationGraph.
Not too hard to fix, in theory.

**Issue 2: Parameter initialization**
Parameter initialization is done on a per-layer basis. Probably not the end of the world if we initialize shared parameters multiple times, though it is of course a bit inefficient to do so; I'd prefer an "initialize shared parameters only in first layer to be created with them". Parameter initialization also depends on fan-in/fan-out in many cases. Not sure yet if that aspect of weight init should be modified by parameter sharing, but we'll need to think about it.


**Issue 3: Updaters and updater blocks**
Two sub-issues here. First, there's an issue where you have different updater configuration for layers with shared parameters. We'll need to enforce same updater config for both. Can't have different learning rate on 2 different layers for shared parameters, for example.

Second, we currently implement a scheme where we combine updaters across parameters with same updater configuration, that are contiguous in the updater array. So instead of doing adamUpdater(W0), adamUpdater(b0), adamUpdater(W1), etc in discrete steps, we instead do adamUpdater([W0,b0,W1,...]).
What does this mean in practice? We'll need to exclude layers when the parameters for that layer have already been seen.
That will haven in BaseMultiLayerUpdater and UpdaterBlock:
https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/updater



**Issue 4: Layers are implemented to overwrite the gradient view array contents during backprop.**

Example: this gemm call for dense/output layers:
https://github.com/deeplearning4j/deeplearning4j/blob/b3345bc2cae706542a9fdbd16ef3e69430f24af1/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/BaseLayer.java#L108
```Nd4j.gemm(input.castTo(weightGrad.dataType()), delta, weightGrad, true, false, 1.0, 0.0); ```
That's weightGrad = 0.0 * weightGrad + 1.0 * input x delta.

In the case of parameter sharing, we need the following behaviour:
(a) For first layer using a given set of shared parameters during backprop, we need the existing implementation
(b) For all subsequent layers, we need ```Nd4j.gemm(..., 1.0, 1.0)```
Somehow, the layer needs to know which case it should implement.

So if layers A and B have shared parameters, and topological order is A,...,B then B should be using 0.0 in gemm (i.e., =) in backprop whereas A should be using 1.0 (i.e., +=).

We'll need changes for this in all layers with parameters.

**Comment by rnett** [comment]
Yeah, that's quite a bit more involved than I thought it was.

From what I can tell, the place parameters get updated is [here](https://github.com/deeplearning4j/deeplearning4j/blob/e50ece1ade7433946c1438bfe0713d2da092f6a9/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/BaseOptimizer.java#L218), and the whole updater setup is for updating gradients.  Is that correct?

**Comment by rnett** [comment]
Alright, this is what I'm looking at as far as a potential implementation:

Store weights and gradients in blocks, multiple unshared params can be in the same block.  Each block is a contiguous column vector.  Have a method to create a single row vector for the whole thing, but it won't be a view.  Updater blocks and layer views are subsets of these blocks.

Depending on how many layers are shared, this could hurt performance quite a bit.  If no layers are shared, it should default to the current implementation automatically.

Theoretically, if shared layers are used in the same order in multiple models (Assuming there isn't some unforeseen issue there), you could put them in the same block, as long as they are all distinct (not shared amongst themselves).

For parameters:
 * Loading.  Split the array and set appropriate blocks, which is expensive, but its a one time cost.
 * Parameter initialization.  Initialization is done by layer view, so we can get the view from the correct block.  
 * Updating: no changes needed, already calls `setParams` method.  Will require splitting the array up properly.  The better option is to use blocks in `step()`, which shouldn't be too hard.

For gradients:

All updates are done either on layer views or on updater blocks.  Within each of those, we can provide a view, regardless of whether the layer is shared or not.

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
